AMQP: add more options to configuration of the ConnectionFactory, #191

matsluni opened the pull request:
Hi,

I added additional configuration options as described in #191. 

Please let me know what you think.

Matthias

2m commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnector.scala:
```diff
@@ -15,19 +15,42 @@ private[amqp] trait AmqpConnector {
+        if (hostAndport.size == 1) {
+          factory.setHost(hostAndport.head._1)
+          factory.setPort(hostAndport.head._2)
+        }
```
If more than one host and port provided, then no host nor port will be set on factory?

matsluni replied:
Yes, if there are more than 1 entries, the newConnection() method in the AmqpConnector will be used, where all the entries from the list can be given to create a newConnection.

Or do you think it would be better to set at least one host and port here, just in case....?

2m replied:
Alright, now I understand the idea. However I still see some discrepancy between the way `hostAndPortList` is handled here and in `newConnection`. 

Here it checks if the size is one. And in newConnection it checks if the list is non empty. So if the hostAndPortList has only one element the host and port will be set here as well as passed to the `factory.newConnection` method. But if the hostAndPortList has more than one element then these will be passed only when calling `factory.newConnection`. 

Can we remove the `factory.setHost` and `factory.setPort` altogether?

2m commented on amqp/src/main/scala/akka/stream/alpakka/amqp/model.scala:
```diff
@@ -121,15 +121,25 @@ object AmqpConnectionUri {
+    networkRecoveryInterval: Option[Int] = None,
+    automaticRecoveryEnabled: Option[Boolean] = None,
+    topologyRecoveryEnabled: Option[Boolean] = None,
+    exceptionHandler: Option[ExceptionHandler] = None
```
We have to provide a way to set these from Java, which means adding `with***` methods for every one of these properties.

matsluni replied:
Ok, I added the `with***` methods.

matsluni replied:
I also added a `addHostAndPort(hostAndPorts: (String, Int))` and the Java one `addHostAndPort(hostAndPort: akka.japi.Pair[String, Int])` to add host/port combinations to the already given list. I am a bit unsure about the name `addHostAndPort`. What do you think?

2m replied:
Looking at other connectors (amqp and mqtt) the convention is still to use `with***` method name, but also make it a vararg method, like so:

```scala
@annotation.varargs
def withHostAndPorts(hostAndPorts: akka.japi.Pair[String, Int]*) = ...
```

2m commented:
@matsluni I have pushed a commit with a formatting fix so CI could run. After a couple of nitpick fixes this can go in.

matsluni commented:
I was on vacation, but now I am back. I will have a look at it tomorrow.

matsluni commented:
I have updated the PR with the comments from above, which mainly are:

- stick to the `with***` pattern 
- also removed `setHost` and `setPort` on the `factory` as these are now arguments to `factory.newConnection`

I hope this is good now. Thank you very much for your patience.

Best regards,
Matthias

matsluni commented:
hmmm.....travis build failed due to formatting error. I did run `test:compile` again but this didn't changed the formatting to the one travis log showed. Any ideas?

2m commented:
I see. Formatting rules have changed a tiny bit in master and Travis builds PRs after a merging them onto current master giving unformatted code error.

Rebase this PR onto master and then `test:compile` will reformat code according to the latest rules. After committing these formatting changes it should be all good.

matsluni commented:
I rebased the branch onto current master. Looks better now. :-)

2m commented:
Good stuff. Thanks a lot. :)
